src/librbd/api/PoolMetadata.cc

```diff
@@ -0,0 +1,106 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/api/PoolMetadata.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/Utils.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::PoolMetadata: " << __func__ << ": "
+
+namespace librbd {
+namespace api {
+
+template <typename I>
+int PoolMetadata<I>::get(librados::IoCtx& io_ctx,
+                     const std::string &key, std::string *value) {
+  CephContext *cct = (CephContext *)io_ctx.cct();
+
+  int r = cls_client::metadata_get(&io_ctx, RBD_INFO, key, value);
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed reading metadata " << key << ": " << cpp_strerror(r)
+               << dendl;
+  }
+
+  return r;
+}
+
+template <typename I>
+int PoolMetadata<I>::set(librados::IoCtx& io_ctx, const std::string &key,
+                         const std::string &value) {
+  CephContext *cct = (CephContext *)io_ctx.cct();
+
+  std::string config_key;
+  if (util::is_metadata_config_override(key, &config_key)) {
+    int r = ConfigProxy{false}.set_val(config_key.c_str(), value);
+    if (r < 0) {
+      lderr(cct) << "validation for " << key << " failed: " << cpp_strerror(r)
+                 << dendl;
+      return -EINVAL;
+    }
+  }
+
+  ceph::bufferlist bl;
+  bl.append(value);
+
+  int r = cls_client::metadata_set(&io_ctx, RBD_INFO, {{key, bl}});
+  if (r < 0) {
+    lderr(cct) << "failed setting metadata " << key << ": " << cpp_strerror(r)
+               << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int PoolMetadata<I>::remove(librados::IoCtx& io_ctx, const std::string &key) {
+  CephContext *cct = (CephContext *)io_ctx.cct();
+
+  std::string value;
+  int r = cls_client::metadata_get(&io_ctx, RBD_INFO, key, &value);
+  if (r < 0) {
+    if (r == -ENOENT) {
+      lderr(cct) << "metadata " << key << " does not exist" << dendl;
```

Comment:
I was trying to preserve the same behavior that we have for image metadata, and there you decided it was important to return `-ENOENT`. I  would like if `metadata_remove` checked and returned `-ENOEN`  by the osd, but for image metadata  we did not choose this approach then. Probably because it would not work for older osds?

If you think it is ok to change `metadata_remove` cls method I can do this.

[1] http://tracker.ceph.com/issues/16990

Reply:
OK, if the ```metadata_remove``` cls method doesn't return ```-ENOENT```, that's fine then. Yeah, it's a backwards compatibility thing since it would most likely break the upgrade tests if the behavior was changed.